README.md
higlo
=====

Syntax highlighter in OCaml.

More information at http://zoggy.github.io/higlo/ .


Changes
Next release:
- fix highlight of %id to handle capitalized ids and ids with dots

Release 0.6:
- OCaml lexer: use same keyword level for %lwt as for other %id
- OCaml lexer: fix highlight of %id (% was removed except for %lwt)

Release 0.5:
- add json and dot lexers
- support "nonrec" keyword in ocaml lexer
- use Xtmpl >= 0.13
- OCaml lexer: handle %id as Keyword 5

Release 0.4:
- safe-string mode
- OCaml lexer: fix #2: bad lexing of directives on first line
- OCaml lexer: handle Lwt keywords as Keyword 10

Release 0.3:
- new higlo tool
- new mk-higlo script

Release 0.2:
- ocaml lexer: handle toplevel directives
- META: add separate packages higlo.ocaml and higlo.xml,
  required by higlo.lexers



higlo_test.ml
(*********************************************************************************)
(*                Higlo                                                          *)
(*                                                                               *)
(*    Copyright (C) 2014 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

(** *)

(*c==v=[File.string_of_file]=1.0====*)
let string_of_file name =
  let chanin = open_in_bin name in
  let len = 1024 in
  let s = String.create len in
  let buf = Buffer.create len in
  let rec iter () =
    try
      let n = input chanin s 0 len in
      if n = 0 then
        ()
      else
        (
         Buffer.add_substring buf s 0 n;
         iter ()
        )
    with
      End_of_file -> ()
  in
  iter ();
  close_in chanin;
  Buffer.contents buf
(*/c==v=[File.string_of_file]=1.0====*)


let files = ref [];;
let lang = ref "ocaml";;

type mode = Tokens | Xtmpl | Html
let mode = ref Xtmpl

let options = [
    "--tokens", Arg.Unit (fun () -> mode := Tokens), " Output tokens only" ;
    "--html", Arg.Unit (fun () -> mode := Html), " Output an HTML page" ;
    "--lang", Arg.Set_string lang, "<s> set language to <s>; default is ocaml" ;
  ]
;;

let handle_file file =
  match !mode with
    Tokens ->
      let tokens = Higlo.parse ~lang: !lang (string_of_file file) in
      List.iter (fun t -> print_endline (Higlo.string_of_token t)) tokens
  | Xtmpl ->
      let xmls = Higlo.to_xtmpl ~lang: !lang (string_of_file file) in
      print_endline (Xtmpl.string_of_xmls xmls)
  | Html ->
      let xmls = Higlo.to_xtmpl ~lang: !lang (string_of_file file) in
      print_string "<html><head>
      <meta content=\"text/html; charset=utf-8\" http-equiv=\"Content-Type\"/>
      <link href=\"style.css\" rel=\"stylesheet\" type=\"text/css\"/>
      </head>
      <body><pre>";
      print_string (Xtmpl.string_of_xmls xmls);
      print_string "</pre></body></html>"
;;

let () =
  try
    Arg.parse options
      (fun f  -> files := f :: !files)
      (Printf.sprintf "Usage: %s [options]\nwhere options are:" Sys.argv.(0));
    List.iter handle_file (List.rev !files)
  with
    Failure s ->
      prerr_endline s ;
      exit 1
;;
    

higlo.ml
(*********************************************************************************)
(*                Higlo                                                          *)
(*                                                                               *)
(*    Copyright (C) 2014 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

(** *)

type token =
| Bcomment of string (** block comment *)
| Constant of string
| Directive of string
| Escape of string (** Escape sequence like [\123] *)
| Id of string
| Keyword of int * string
| Lcomment of string (** one line comment *)
| Numeric of string
| String of string
| Symbol of int * string
| Text of string (** Used for everything else *)

let string_of_token = function
| Bcomment s -> Printf.sprintf "Bcomment(%S)" s
| Constant s -> Printf.sprintf "Constant(%S)" s
| Directive s -> Printf.sprintf "Directive(%S)" s
| Escape s -> Printf.sprintf "Escape(%S)" s
| Id s -> Printf.sprintf "Id(%S)" s
| Keyword (n, s) -> Printf.sprintf "Keyword(%d, %S)" n s
| Lcomment s -> Printf.sprintf "Lcomment(%S)" s
| Numeric s -> Printf.sprintf "Numeric(%S)" s
| String s -> Printf.sprintf "String(%S)" s
| Symbol (n, s) -> Printf.sprintf "Symbol(%d, %S)" n s
| Text s -> Printf.sprintf "Text(%S)" s

module Smap = Map.Make (String)
exception Unknown_lang of string

type lexer = Ulexing.lexbuf -> token list

let langs = ref Smap.empty

let get_lexer lang =
  try Smap.find lang !langs
  with Not_found -> raise (Unknown_lang lang)
;;
let register_lang name f = langs := Smap.add name f !langs ;;

let parse ~lang s =
  try
    let lexer = get_lexer lang in
    let lexbuf = Ulexing.from_utf8_string s in
    let len = Array.length (Ulexing.get_buf lexbuf) in
    let b = Buffer.create 256 in
    let rec add_tokens acc = function
      Text s -> Buffer.add_string b s ; acc
    | t ->
        let acc =
          if Buffer.length b > 0 then
            (
             let acc = (Text (Buffer.contents b)) :: acc in
             Buffer.reset b ;
             acc
            )
          else acc
        in
        t :: acc
    in
    let rec iter acc =
      if Ulexing.get_pos lexbuf >= len then
        begin
          let acc =
            if Buffer.length b > 0
            then (Text (Buffer.contents b)) :: acc
            else acc
          in
          List.rev acc
        end
      else
        begin
          let tokens = lexer lexbuf in
          iter (List.fold_left add_tokens acc tokens)
        end
    in
    try iter []
    with Ulexing.Error ->
        let pos = Ulexing.get_pos lexbuf in
        let msg = Printf.sprintf "Lexing error at character %d" pos in
        prerr_endline msg ;
        [Text s]
  with
    Unknown_lang _ -> [Text s]
;;

type classes =
  {
    bcomment : string ;
    constant : string ;
    directive : string ;
    escape : string ;
    id : string ;
    keyword : int -> string ;
    lcomment : string ;
    numeric : string ;
    string : string ;
    symbol : int -> string ;
    text : string ;
  }

let default_classes = {
    bcomment = "comment" ;
    constant = "constant" ;
    directive = "directive" ;
    escape = "escape" ;
    id = "id" ;
    keyword = (function 0 -> "kw" | n -> "kw"^(string_of_int n)) ;
    lcomment = "comment" ;
    numeric = "numeric" ;
    string = "string" ;
    symbol = (function 0 -> "sym" | n -> "sym"^(string_of_int n)) ;
    text = "text" ;
  }
;;

let token_to node =
  fun ?(classes=default_classes) ->
    function
    | Bcomment s -> node classes.bcomment s
    | Constant s -> node classes.constant s
    | Directive s -> node classes.directive s
    | Escape s -> node classes.escape s
    | Id s -> node classes.id s
    | Keyword (n, s) -> node (classes.keyword n) s
    | Lcomment s -> node classes.lcomment s
    | Numeric s -> node classes.numeric s
    | String s -> node classes.string s
    | Symbol (n, s) -> node (classes.symbol n) s
    | Text s -> node classes.text s

let token_to_xml =
  let module X = Xtmpl_xml in
  let node cl cdata =
    let atts = X.atts_one ("","class") (cl, None) in
    X.node ("","span") ~atts [X.cdata cdata]
  in
  token_to node
;;

let token_to_xml_rewrite =
  let module X = Xtmpl_rewrite in
  let node cl cdata =
    let atts = X.atts_one ("","class") [X.cdata cl] in
    X.node ("","span") ~atts [X.cdata cdata]
  in
  token_to node
;;

let to_xml ?classes ~lang s =
  List.map (token_to_xml ?classes) (parse ~lang s)
;;

let to_xml_rewrite ?classes ~lang s =
  List.map (token_to_xml_rewrite ?classes) (parse ~lang s)
;;



higlo_printers.ml
(*********************************************************************************)
(*                Higlo                                                          *)
(*                                                                               *)
(*    Copyright (C) 2014 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

(** *)

type printer = Higlo.token list -> unit

module SMap = Map.Make(String);;

let printers = ref SMap.empty ;;

let get_printer name =
  try SMap.find name !printers
  with Not_found -> failwith (Printf.sprintf "Unknown printer %S" name)
;;

let register_printer name f = printers := SMap.add name f !printers;;

let xml_printer tokens =
  let xmls = List.map Higlo.token_to_xml tokens in
  print_string (Xtmpl_xml.to_string xmls)
;;

let html_printer tokens =
  print_string "<html>
  <head>
    <meta content=\"text/html; charset=utf-8\" http-equiv=\"Content-Type\"/>
    <link href=\"style.css\" rel=\"stylesheet\" type=\"text/css\"/>
  </head>
  <body><pre>";
  xml_printer tokens;
  print_string "</pre></body></html>"
;;

let token_printer tokens =
  List.iter (fun t -> print_string (Higlo.string_of_token t)) tokens
;;

let () =
  List.iter (fun (name, f) -> register_printer name f)
    [ "xml", xml_printer ;
      "html", html_printer ;
      "tokens", token_printer ;
    ]
;;


higlo_main.ml
(*********************************************************************************)
(*                Higlo                                                          *)
(*                                                                               *)
(*    Copyright (C) 2014 Institut National de Recherche en Informatique          *)
(*    et en Automatique. All rights reserved.                                    *)
(*                                                                               *)
(*    This program is free software; you can redistribute it and/or modify       *)
(*    it under the terms of the GNU Lesser General Public License version        *)
(*    3 as published by the Free Software Foundation.                            *)
(*                                                                               *)
(*    This program is distributed in the hope that it will be useful,            *)
(*    but WITHOUT ANY WARRANTY; without even the implied warranty of             *)
(*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *)
(*    GNU Library General Public License for more details.                       *)
(*                                                                               *)
(*    You should have received a copy of the GNU Lesser General Public           *)
(*    License along with this program; if not, write to the Free Software        *)
(*    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA                   *)
(*    02111-1307  USA                                                            *)
(*                                                                               *)
(*    Contact: Maxence.Guesdon@inria.fr                                          *)
(*                                                                               *)
(*                                                                               *)
(*********************************************************************************)

(** *)

type to_load = Pkgs of string list | Files of string list

let verbose = ref false;;
let verb msg = if !verbose then prerr_endline msg;;

let string_of_inch chanin =
  let len = 1024 in
  let s = Bytes.create len in
  let buf = Buffer.create len in
  let rec iter () =
    try
      let n = input chanin s 0 len in
      if n = 0 then
        ()
      else
        (
         Buffer.add_subbytes buf s 0 n;
         iter ()
        )
    with
      End_of_file -> ()
  in
  iter ();
  Buffer.contents buf
;;

let string_of_file name =
  let chanin = open_in_bin name in
  let s = string_of_inch chanin in
  close_in chanin ;
  s
;;

let handle_file lang printer file =
  let tokens = Higlo.parse ~lang (string_of_file file) in
  printer tokens
;;

let handle_stdin lang printer =
  let code = string_of_inch stdin in
  let tokens = Higlo.parse ~lang code in
  printer tokens
;;

(*c==v=[String.split_string]=1.2====*)
let split_string ?(keep_empty=false) s chars =
  let len = String.length s in
  let rec iter acc pos =
    if pos >= len then
      match acc with
        "" -> if keep_empty then [""] else []
      | _ -> [acc]
    else
      if List.mem s.[pos] chars then
        match acc with
          "" ->
            if keep_empty then
              "" :: iter "" (pos + 1)
            else
              iter "" (pos + 1)
        | _ -> acc :: (iter "" (pos + 1))
      else
        iter (Printf.sprintf "%s%c" acc s.[pos]) (pos + 1)
  in
  iter "" 0
(*/c==v=[String.split_string]=1.2====*)

(*c==v=[List.list_remove_doubles]=1.0====*)
let list_remove_doubles ?(pred=(=)) l =
  List.fold_left
    (fun acc e -> if List.exists (pred e) acc then acc else e :: acc)
    []
    (List.rev l)
(*/c==v=[List.list_remove_doubles]=1.0====*)

let load_file file =
  let file = Dynlink.adapt_filename file in
  verb (Printf.sprintf "Loading file %s" file);
  try Dynlink.loadfile file
  with Dynlink.Error e ->
      failwith (Dynlink.error_message e)

let files_of_packages pkg_names =
  let kind = if Dynlink.is_native then "native" else "byte" in
  let tmp_file = Filename.temp_file "higlo" ".txt" in
  let com =
    Printf.sprintf "ocamlfind query %s -predicates plugin,%s -r -format %%d/%%a > %s"
      (String.concat " " (List.map Filename.quote pkg_names))
      kind (Filename.quote tmp_file)
  in
  match Sys.command com with
    0 ->
      let s = string_of_file tmp_file in
      Sys.remove tmp_file ;
      split_string s ['\n']
  | n ->
      Sys.remove tmp_file ;
      let msg = Printf.sprintf "Command failed (%d): %s" n com in
      failwith msg
;;


let files_of_to_load = function
  Pkgs pkgs -> files_of_packages pkgs
| Files files -> files
;;

let dynload_code l =
  let files = List.fold_left
    (fun acc to_load ->
       List.fold_left
         (fun acc f -> if List.mem f acc then acc else f :: acc)
         acc (files_of_to_load to_load)
    )
    [] l
  in
  List.iter load_file (List.rev files)
;;

let to_load = ref [];;

let add_pkgs s = to_load := (Pkgs (split_string s [','; ' '])) :: !to_load ;;
let add_dynf s = to_load := (Files (split_string s [','; ' '])) :: !to_load;;

let files = ref [];;
let lang = ref "ocaml";;
let out_format = ref "xml";;

let options = [
    "-v", Arg.Set verbose, " verbose mode";

    "-f", Arg.Set_string out_format,
    "<s> set output format to <s>; default is xml" ;

    "-l", Arg.Set_string lang, "<s> set language to <s>; default is ocaml" ;

    "--pkg", Arg.String add_pkgs,
    "pkg1,pkg2,... dynmically load the given packages" ;

    "--load", Arg.String add_dynf,
    "file1.cm[xs|o|a],... dynmically load the given object files" ;
  ]
;;

let () =
  try
    Arg.parse (Arg.align options)
      (fun f  -> files := f :: !files)
      (Printf.sprintf "Usage: %s [options] [files]\nwhere options are:" Sys.argv.(0));
    dynload_code (List.rev !to_load);
    let printer = Higlo_printers.get_printer !out_format in
    ignore(
     try let _x = Higlo.get_lexer !lang in ()
     with Higlo.Unknown_lang name -> failwith (Printf.sprintf "Unknown language %S" name)
    );
    match List.rev !files with
      [] -> handle_stdin !lang printer
    | files -> List.iter (handle_file !lang printer) files
  with
    Failure s ->
      prerr_endline s ;
      exit 1
;;
    